src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs

```diff
@@ -1039,50 +1039,82 @@ private void RemoveWorseMembers<TMember>(ArrayBuilder<MemberResolutionResult<TMe
 
             var worse = ArrayBuilder<int>.GetInstance(results.Count, unknown);
 
-            for (int c1Idx = 0; c1Idx < results.Count; c1Idx++)
+            int currentBestIndex;
+            for (currentBestIndex = 0; currentBestIndex < results.Count; currentBestIndex++)
             {
-                var c1result = results[c1Idx];
+                if (results[currentBestIndex].Result.IsValid)
+                {
+                    worse[currentBestIndex] = betterThanEverything;
+                    break;
+                }
+            }
 
-                // If we already know this is worse than something else, no need to check again.
-                if (!c1result.Result.IsValid || worse[c1Idx] == worseThanSomething)
+            for (int index = currentBestIndex + 1; index < results.Count; index++)
+            {
+                if (!results[index].Result.IsValid)
                 {
                     continue;
                 }
 
-                bool c1IsBetterThanEverything = true;
-                for (int c2Idx = 0; c2Idx < results.Count; c2Idx++)
+                var better = BetterFunctionMember(results[currentBestIndex], results[index], arguments.Arguments, ref useSiteDiagnostics);
+                if (better == BetterResult.Left)
+                {
+                    worse[index] = worseThanSomething;
+                }
+                else if (better == BetterResult.Right)
+                {
+                    worse[currentBestIndex] = worseThanSomething;
+                    worse[index] = betterThanEverything;
+                    currentBestIndex = index;
+                }
+                else
                 {
-                    var c2result = results[c2Idx];
+                    worse[currentBestIndex] = notBetterThanEverything;
+                    worse[index] = notBetterThanEverything;
 
-                    if (!c2result.Result.IsValid)
+                    while (++currentBestIndex < results.Count)
                     {
-                        continue;
+                        if (results[currentBestIndex].Result.IsValid)
+                        {
+                            worse[currentBestIndex] = betterThanEverything;
+                            break;
+                        }
                     }
 
-                    if (c1result.Member == c2result.Member)
```

Comment:
> which supposedly won't happen (?)

It feels like we should be absolutely certain about this. The check could be performed with indexes before, but it wasn't. Was that just an oversight, or could this make a difference?